kaldb/src/test/java/com/slack/kaldb/com/slack/kaldb/logstore/search/KaldbDistributedQueryServiceTest.java

```diff
@@ -0,0 +1,293 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class KaldbDistributedQueryServiceTest {
+
+  private SimpleMeterRegistry metricsRegistry;
+
+  private MetadataStore zkMetadataStore;
+  private SearchMetadataStore searchMetadataStore;
+  private SnapshotMetadataStore snapshotMetadataStore;
+  private TestingServer testZKServer;
+  private SearchContext indexer1SearchContext;
+  private SearchContext indexer2SearchContext;
+  private SearchContext cache1SearchContext;
+  private SearchContext cache2SearchContext;
+
+  @Before
+  public void setUp() throws Exception {
+    Tracing.newBuilder().build();
+
+    metricsRegistry = new SimpleMeterRegistry();
+    testZKServer = new TestingServer();
+
+    // Metadata store
+    KaldbConfigs.ZookeeperConfig zkConfig =
+        KaldbConfigs.ZookeeperConfig.newBuilder()
+            .setZkConnectString(testZKServer.getConnectString())
+            .setZkPathPrefix("indexerTest")
+            .setZkSessionTimeoutMs(1000)
+            .setZkConnectionTimeoutMs(1000)
+            .setSleepBetweenRetriesMs(1000)
+            .build();
+
+    zkMetadataStore = spy(ZookeeperMetadataStoreImpl.fromConfig(metricsRegistry, zkConfig));
+
+    snapshotMetadataStore = spy(new SnapshotMetadataStore(zkMetadataStore, true));
+    searchMetadataStore = spy(new SearchMetadataStore(zkMetadataStore, true));
+    indexer1SearchContext = new SearchContext("indexer_host1", 10000);
+    indexer2SearchContext = new SearchContext("indexer_host2", 10001);
+    cache1SearchContext = new SearchContext("cache_host1", 20000);
+    cache2SearchContext = new SearchContext("cache_host2", 20001);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    snapshotMetadataStore.close();
+    searchMetadataStore.close();
+    zkMetadataStore.close();
+    metricsRegistry.close();
+    testZKServer.close();
+  }
+
+  @Test
+  public void testOneIndexer() {
+    Instant chunkCreationTime = Instant.ofEpochMilli(100);
+    Instant chunkEndTime = Instant.ofEpochMilli(200);
+    createIndexerZKMetadata(chunkCreationTime, chunkEndTime, "1", indexer1SearchContext);
+    await().until(() -> snapshotMetadataStore.listSync().size() == 1);
+    await().until(() -> searchMetadataStore.listSync().size() == 1);
+
+    Collection<String> snapshots =
+        getSnapshotsToSearch(
+            snapshotMetadataStore,
+            searchMetadataStore,
+            chunkCreationTime.toEpochMilli(),
+            chunkEndTime.toEpochMilli(),
+            Set.of("1"));
+    assertThat(snapshots.size()).isEqualTo(1);
+    assertThat(snapshots.iterator().next()).isEqualTo(indexer1SearchContext.toString());
+
+    snapshots =
+        getSnapshotsToSearch(
+            snapshotMetadataStore,
+            searchMetadataStore,
+            chunkEndTime.toEpochMilli() + 1,
+            chunkEndTime.toEpochMilli() + 100,
+            Set.of("1"));
+    assertThat(snapshots.size()).isEqualTo(0);
```

Comment:
In addition to size, we should also check the data in the snapshot to make sure that the data we get is the one we expect for all the unit tests.